From patchwork Thu Aug  9 00:45:24 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Junaid Shahid <junaids@google.com>
X-Patchwork-Id: 10560675
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 4BD1414E5
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Aug 2018 00:45:54 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 338B929EA2
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Aug 2018 00:45:54 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 2368129F80; Thu,  9 Aug 2018 00:45:54 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AACE429EA2
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Aug 2018 00:45:53 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1731887AbeHIDHq (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 8 Aug 2018 23:07:46 -0400
Received: from mail-qt0-f202.google.com ([209.85.216.202]:45327 "EHLO
        mail-qt0-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727530AbeHIDHq (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 8 Aug 2018 23:07:46 -0400
Received: by mail-qt0-f202.google.com with SMTP id d14-v6so3192415qtn.12
        for <kvm@vger.kernel.org>; Wed, 08 Aug 2018 17:45:38 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=NvFbrXrZMo/EDGjm42ohxkQY2Oi3bQ9BUbVfQ0c85UM=;
        b=chtBUhGOOqTnw/GM5eMO9JJHlYRZpWYUz6HvduGEx8ugphuVza8b0o3xjM0C738Oyw
         08GEAB9+VsEzrlCp3y8ezT4yCSh+l+t6wIO0TJDGltLuzODp8fbzRKSzovMf+IWWYVn8
         FMlgFcMXipcI0ZAgWVKNS/0K1SoXdXVMZst5x2VYM0oiZ4anxP4D/cCHnSeFK5+UxC58
         u1kwqvb6Eb3FCXtHMvMI1tNnq76wKa70H+beYGRY7WZg5MqZrQuFS6Re9SmrPpgJHAXu
         yI55+I6gWsI+FntAVLtsRFjt+9PVx8IUXS0OJxahuAdjzT4AF3Zw/isNVuIwgaEV2TlM
         dGIw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=NvFbrXrZMo/EDGjm42ohxkQY2Oi3bQ9BUbVfQ0c85UM=;
        b=SB/DF5GMgnOy+/FbP/FFzZfbPsVIkqIOjloPertmTiWb40bN5hb1PlcGFDBby59n0n
         BOJZS5vETaF2jlJpvC9DiTdRyJzSQMewLXy0LjwwpTxBFKGK+6laIy5PfrS9pyVY4OA7
         fGPPwTRfiuzsLX19PnlDBpVzjtylEplhS+3JOuFhujGLRzG2esDM27KO3KEGWs0H4mxM
         T7TKNHcrEDY5lbNUGp4tOMSHtDd9PI5GghlNfR70mbD/QB2NoiFNtzyc3Ing+LQ/3mhP
         V3vFHjLu8AuNlgajk2qCGilpgGNFvH8DlIAwUdyYq/eFRTDuRutlE/bfY9hwm2gDOL09
         YGKg==
X-Gm-Message-State: AOUpUlH89s80asOGTqlAWo27weoFxbriL4AlhynblMMgd2UNKrPQF1EE
        PD76BbhaD0VrqDl1gaQuO3lc8jChS2VV
X-Google-Smtp-Source: 
 AA+uWPxo7/MMmYZovk5f3PYwWkZIUKSlvMDDAV2ujMEbqqdwgnkBHaY7T73OsgLrJqQicN+qw5ZDTlyKQM68
X-Received: by 2002:a37:de0e:: with SMTP id
 h14-v6mr168390qkj.50.1533775538387;
 Wed, 08 Aug 2018 17:45:38 -0700 (PDT)
Date: Wed,  8 Aug 2018 17:45:24 -0700
Message-Id: <20180809004524.129883-1-junaids@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.18.0.345.g5c9ce644c3-goog
Subject: [PATCH] kvm: mmu: Don't read PDPTEs when paging is not enabled
From: Junaid Shahid <junaids@google.com>
To: pbonzini@redhat.com
Cc: kvm@vger.kernel.org, andreslc@google.com, jmattson@google.com
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

kvm should not attempt to read guest PDPTEs when CR0.PG = 0 and
CR4.PAE = 1.

Signed-off-by: Junaid Shahid <junaids@google.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/kvm/x86.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 3c83711c0ebe..a726af7d31b6 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -627,7 +627,7 @@ bool pdptrs_changed(struct kvm_vcpu *vcpu)
 	gfn_t gfn;
 	int r;
 
-	if (is_long_mode(vcpu) || !is_pae(vcpu))
+	if (is_long_mode(vcpu) || !is_pae(vcpu) || !is_paging(vcpu))
 		return false;
 
 	if (!test_bit(VCPU_EXREG_PDPTR,
@@ -8123,7 +8123,7 @@ static int __set_sregs(struct kvm_vcpu *vcpu, struct kvm_sregs *sregs)
 		kvm_update_cpuid(vcpu);
 
 	idx = srcu_read_lock(&vcpu->kvm->srcu);
-	if (!is_long_mode(vcpu) && is_pae(vcpu)) {
+	if (!is_long_mode(vcpu) && is_pae(vcpu) && is_paging(vcpu)) {
 		load_pdptrs(vcpu, vcpu->arch.walk_mmu, kvm_read_cr3(vcpu));
 		mmu_reset_needed = 1;
 	}
